ObjectCapabilities/revokableForwarder.rho

```diff
@@ -31,14 +39,67 @@ new revokFactory in {
        * Messages to be forwarded are sent here. If the forwarder
        * if still active the message is forwarded, and an ack is
        * sent. If the forwarder is inactive, no forwarding happens
-       * and no ack is sent.
+       * and no ack is sent. The ack is just the forwarder saying,
+       * Your message got here and I passed it on. It is _not_ a
+       * guarantee that your message reaches an ultimate downstream
+       * capability in the case that several forwarders are composed.
        */
       contract forward(@message, ack) = {
         // for (_ <! activeCh) {
         for (_ <- activeCh) {
           activeCh!(Nil) |
+          ack!(Nil) |
           capability!(message)
         }
+      } |
+
+      /**
+       * Overloaded forward method nearly identical to the
+       * previous, but no ack is sent.
+       */
+       contract forward(@message) = {
+         // for (_ <! activeCh) {
+         for (_ <- activeCh) {
+           activeCh!(Nil) |
+           capability!(message)
+         }
+       }
+    }
+  } |
+
+  // Demo of using the forwarder.
+  new superpower,
+    killSwitchCh,
+    forwardCh,
+    ack,
+    stdout(`rho:io:stdout`)
+    in {
+
+    // Silly superpower that we'll wrap with a forwarder
+    // Explicit tupling is not ideal. But I don't see a better option.
+    contract superpower(@(payload /\ String, return)) = {
+      @return!("Superpower Invoked with " ++ payload)
+    } |
+
+    // Create a forwarder for the superpower
+    // revokFactory!(*superpower, await *killSwitchCh, await *forwardCh) {
+    revokFactory!(*superpower, *killSwitchCh, *forwardCh) |
+    for (killSwitch <- killSwitchCh; forward <- forwardCh) {
```

Comment:
Because they'll frequently be going to different places. If I'm granting you temporary access to my superpower, `forward` goes to you (maybe your locker, or some contract you've set up). And `killSwitch` goes to me.

Reply:
> I'm pretty sure a join costs more than pattern matching to split out the parts after returning them on one channel.

Could be. I'm not yet considering performance of any implementation when designing these.